### PR TITLE
Add Trino 389 release notes

### LINKED ISSUE #13043: Release notes for 389

### PR DESCRIPTION
**Pull request by colebow**
## Description

Assemble the release notes for the upcoming Trino 389 release.

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Docs only.

> How would you describe this change to a non-technical end user or system administrator?

Add the release notes for Trino 389 with sections for general, security, connectors and other topics as included in the release.

## Related issues, pull requests, and links

Fixes #13043

## Documentation

( ) No documentation is needed.
(✅) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(✅) No release notes entries required.
( ) Release notes entries required with the following suggested text:

## Verification for each commit

Format: commit, committer/maintainer, PR/issue number, ✅ / ❌ rn ✅ / ❌ docs
✅ rn - release note added and verified, or assessed to be not necessary, set to ❌ rn before completion
✅ docs - need for docs assessed and merged, or assessed to be not necessary, set to ❌ docs before completion

### CODE REVIEW

File: docs/src/main/sphinx/release/release-389.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Release 389 (6 July 2022)
+
+## General
+
+* Improve performance of queries which contain a large number of aggregations
+  that use `RowType` for representing intermediate state (e.g. `SUM`). ({issue}`12762`)
+* Remove deprecated version of `ConnectorRecordSetProvider#getRecordSet`. ({issue}`13084`)
+* Fix query failure when no arguments are passed to a table function. ({issue}`12951`)
+* Fix error where a split keeps running after query failure due to an
+  uninterruptible regular expression function. ({issue}`13064`)
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-389.md:10]: @tangjiangling is this acceptable? I don't know how necessary it is to strictly mention JONI

Comment (tangjiangling) [reply docs/src/main/sphinx/release/release-389.md:10]: I think we should mention JONI, because the issue usually occurs when using the JONI regular library (which can take exponential time for certain regular expressions)

File: docs/src/main/sphinx/release/release-389.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Release 389 (6 July 2022)
+
+## General
+
+* Improve performance of queries which contain a large number of aggregations
+  that use `RowType` for representing intermediate state (e.g. `SUM`). ({issue}`12762`)
```
Comment (ebyhr) [inline docs/src/main/sphinx/release/release-389.md:6]: `RowType` → `row` type

File: docs/src/main/sphinx/release/release-389.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+* Remove deprecated version of `ConnectorRecordSetProvider#getRecordSet`. ({issue}`13084`)
+* Fix query failure when no arguments are passed to a table function. ({issue}`12951`)
+* Fix error where a split keeps running after query failure due to an
+  uninterruptible regular expression function. ({issue}`13064`)
+
+## Delta Lake connector
+
+* Improve native parquet writer performance for flat types. ({issue}`13030`)
```
Comment (ebyhr) [inline docs/src/main/sphinx/release/release-389.md:14]: I suppose users won't understand what "flat types" is.

File: docs/src/main/sphinx/release/release-389.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  the data was written to the table for the first time. ({issue}`12892`)
+
+## PostgreSQL connector
+
+* Prevent creating a new table with a name longer than the max length.
+  Previously, the name was truncated to the max length. ({issue}`12892`)
+
+## SPI
+
+* Fix potential failure when passing default arguments to a table function.
```
Comment (ebyhr) [inline docs/src/main/sphinx/release/release-389.md:40]: nit: Issue number is missing.

File: docs/src/main/sphinx/release/release-389.md (hunk 3 above)
Comment (raunaqmorarka) [reply docs/src/main/sphinx/release/release-389.md:14]: Maybe "primitive" or "non-nested" types then ?

File: docs/src/main/sphinx/release/release-389.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+* Add `raw_query` table function for full query pass-through to the connector. ({issue}`12324`)
+
+## Hive connector
+
+* Improve native parquet writer performance for primitive types. ({issue}`13030`)
+
+## Iceberg connector
+
+* Improve native parquet writer performance for primitive types. ({issue}`13030`)
```
Comment (raunaqmorarka) [inline docs/src/main/sphinx/release/release-389.md:32]: nit: in iceberg native parquet writer is the only writer, so we can drop "native"

File: docs/src/main/sphinx/release/release-389.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+* Fix error where the server does not start if the internal table snapshots
+  cache is disabled. ({issue}`13086`)
+
+## Elasticsearch connector
+
+* Add `raw_query` table function for full query pass-through to the connector. ({issue}`12324`)
+
+## Hive connector
+
+* Improve native parquet writer performance for primitive types. ({issue}`13030`)
```
Comment (raunaqmorarka) [inline docs/src/main/sphinx/release/release-389.md:28]: native -> optimized

File: docs/src/main/sphinx/release/release-389.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
+# Release 389 (6 July 2022)
+
+## General
+
+* Improve performance of queries which contain a large number of aggregations
+  that use the `row` type for representing intermediate state (e.g. `SUM`). ({issue}`12762`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:6]: use "for example"  .. not e.g.

File: docs/src/main/sphinx/release/release-389.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+* Improve join performance when spill to disk is not used. The old behavior for
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:7]: Not sure I understand ... spill to disk was not enabled by default in the past either..

File: docs/src/main/sphinx/release/release-389.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  the spilling join operator can be restored by setting either the
+  `force_spilling_join` configuration property or the
+  `force-spilling-join-operator` session property to false. ({issue}`12618`)
+* Remove deprecated version of `ConnectorRecordSetProvider#getRecordSet`. ({issue}`13084`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:11]: Belongs into SPI section?

File: docs/src/main/sphinx/release/release-389.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+* Fix query failure when no arguments are passed to a table function. ({issue}`12951`)
+* Fix error where a split keeps running after query failure due to an
+  uninterruptible JONI regular expression function. ({issue}`13064`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:14]: JOIN vs JONI?

also is uninterruptible a word in meriam webster?

File: docs/src/main/sphinx/release/release-389.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+## Delta Lake connector
+
+* Improve native parquet writer performance for primitive types. ({issue}`13030`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:18]: Parquet

Comment (mosabua) [reply docs/src/main/sphinx/release/release-389.md:18]: Also ... does that also affect Hive connector and Iceberg connector?

File: docs/src/main/sphinx/release/release-389.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+* Fix error where the server does not start if the internal table snapshots
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:19]: Reword ... maybe

Prevent failing server start when the internal

File: docs/src/main/sphinx/release/release-389.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  cache is disabled. ({issue}`13086`)
+
+## Elasticsearch connector
+
+* Add `raw_query` table function for full query pass-through to the connector. ({issue}`12324`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:24]: Is it really called raw_query ... as compared to query in the JDBC connectors .. if so .. why the inconsistency

File: docs/src/main/sphinx/release/release-389.md (hunk 6 above)
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:28]: Parquet

File: docs/src/main/sphinx/release/release-389.md (hunk 5 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-389.md:32]: Hm ... should we drop it everywhere?

File: docs/src/main/sphinx/release/release-389.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+* Improve native parquet writer performance for primitive types. ({issue}`13030`)
+
+## Iceberg connector
+
+* Improve native parquet writer performance for primitive types. ({issue}`13030`)
+
+## MongoDB connector
+
+* Create a collection when creating a new table. Previously, it was created when
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:36]: Do we need to add a why? or what this improves?

File: docs/src/main/sphinx/release/release-389.md (hunk 5 above)
Comment (raunaqmorarka) [reply docs/src/main/sphinx/release/release-389.md:32]: In other places we call it optimised-writer in the configs

File: docs/src/main/sphinx/release/release-389.md (hunk 8 above)
Comment (raunaqmorarka) [reply docs/src/main/sphinx/release/release-389.md:7]: Currently spill to disk adds some overheads to join operator even if spill to disk is disabled just because the code is more complicated to handle the extra functionality. We're moving towards simplification of join operator by removing the spill to disk related code as this functionality is deprecated in favour of Tardigrade.
However, as an intermediate step we're keeping the old complicated join code which supports spill to disk around which creating a new simplified version of the code without spill to disk support. This is a temporary situation until Tardigrade is more mature.
Also, the new simplified join code will receive further optimisations which were not possible in the old code in near future.
The force-spilling-join-operator config is an escape hatch to allow falling back to old code in case of unforeseen issues.

File: docs/src/main/sphinx/release/release-389.md (hunk 10 above)
Comment (colebow) [reply docs/src/main/sphinx/release/release-389.md:14]: JONI is correct. It's a regex library. And [yes](https://www.merriam-webster.com/dictionary/uninterruptible).

File: docs/src/main/sphinx/release/release-389.md (hunk 11 above)
Comment (colebow) [reply docs/src/main/sphinx/release/release-389.md:18]: @raunaqmorarka - should we include this in those sections as well?

File: docs/src/main/sphinx/release/release-389.md (hunk 13 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-389.md:24]: https://github.com/trinodb/trino/pull/12324#discussion_r908852229

File: docs/src/main/sphinx/release/release-389.md (hunk 14 above)
Comment (colebow) [reply docs/src/main/sphinx/release/release-389.md:36]: I think it's just to better handle long table names, and is a change in behavior which people should be aware of.

File: docs/src/main/sphinx/release/release-389.md (hunk 11 above)
Comment (colebow) [reply docs/src/main/sphinx/release/release-389.md]: Yeah, it very much is. 🤦

File: docs/src/main/sphinx/release/release-389.md (hunk 8 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-389.md:8]: This should just be:
* Improve performance when spilling to disk is disabled.

> The force-spilling-join-operator config is an escape hatch to allow falling back to old code in case of unforeseen issues.

Not just an escape hatch -- it's needed if you do want to run with spilling enabled.

I haven't read that PR, but I don't understand why we need a separate config option vs piggybacking on the options to enable spilling. cc @raunaqmorarka

File: docs/src/main/sphinx/release/release-389.md (hunk 10 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-389.md:14]: I don't think this fixes an error. I would rephrase as:

* Fix potential resource leaks when calling long-running regular expression functions.

File: docs/src/main/sphinx/release/release-389.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  the data was written to the table for the first time. ({issue}`12892`)
+
+## PostgreSQL connector
+
+* Prevent creating a new table with a name longer than the max length.
+  Previously, the name was truncated to the max length. ({issue}`12892`)
+
+## SPI
+
+* Fix potential failure when passing default arguments to a table function. ({issue}`12951`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-389.md:46]: Remove this from the SPI section

File: docs/src/main/sphinx/release/release-389.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+* Fix potential incorrect results for queries involving non-equality filters on
+  top of an outer join. ({issue}`13109`)
+* Fix query failure when no arguments are passed to a table function. ({issue}`12951`)
+* Fix potential resource leaks when calling long-running regular expression
+  functions. ({issue}`13064`)
+
+## Delta Lake connector
+
+* Improve optimized Parquet writer performance for
+  [non-structural data types](https://trino.io/docs/current/language/types.html#structural). ({issue}`13030`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:18]: wrong link syntax

File: docs/src/main/sphinx/release/release-389.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+## MongoDB connector
+
+* Create a collection when creating a new table. Previously, it was created when
+  the data was written to the table for the first time. ({issue}`12892`)
+
+## Phoenix connector
+
+* Add support for JDK version 17. ({issue}`13108`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-389.md:43]: Add support for Java 17 as runtime.

File: docs/src/main/sphinx/release/release-389.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
+# Release 389 (7 July 2022)
+
+## General
+
+* Add support for printing initial plans for queries involving table functions. ({issue}`13106`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-389.md:5]: "Initial plans" is an internal implementation concept. Let's rephrase as:

> * Fix potential failure when using `EXPLAIN` with queries involving table functions.
